liquibase/liquibase
# Make EnterpriseDBDatabase detection less broad

## Description
The current implementation of EnterpriseDBDatabase picks up database with just "edb" in the url anywhere vs. an actual edb url.

Fixes #2363 

## Work Around
You can manually override the (wrongly) detected dialect with `databaseClass: liquibase.database.core.MySQLDatabase` or `databaseClass: liquibase.database.core.DB2Database` etc. in your liquibase.properties file.

- - -
## Dev Handoff Notes (Internal Use)
#### Links
* Fixed Issue: [https://github.com/liquibase/liquibase/issues/2363](https://github.com/liquibase/liquibase/issues/2363)
* Test Results: [https://github.com/liquibase/liquibase/pull/2364/checks](https://github.com/liquibase/liquibase/pull/2364/checks)

#### Testing
* Steps to Reproduce: 
  * [https://github.com/liquibase/liquibase/issues/2363](https://github.com/liquibase/liquibase/issues/2363)
  * Connect to a non-postgresql url with "edb" in the url somewhere. Like jdb:mysql:localhost/testedb
* Guidance:
  * Impact: Fixes all databases, but change is only in edb code

#### Dev Verification
Added automated test to catch the problem

## Test Requirements (Liquibase Internal QA)
**Manual Tests**

**Setup: Case 1**

* Create a MySQL database named test*edb*1
* Update liquibase.properties file to connect to that database:
`url:jdbc:mysql://localhost:3309/testedb1?useSSL=false&allowPublicKeyRetrieval=true`
* Comment out or delete the `driver` name in liquibase.properties
* Comment out or delete the MySQL driver specified in the liquibase.properties `classpath.`
* Copy the MySQL driver into the liquibase/lib directory 

:warning: I found that if the driver was specified in the liquibase.properties file, you do not encounter this bug, even when a database URL contains the literal string `edb`. That is why the setup removes any reference to the driver class or driver jar. Liquibase does not ship with a MySQL jar in lib; we must add that manually to run this test. :warning:

##### Error Message output By Bug
```
[2022-01-19 12:26:29] INFO [liquibase.database] Cannot check pg_settings
java.sql.SQLSyntaxErrorException: Table 'testedb1.pg_settings' doesn't exist
```

**Verify non-Postgres/EDB JDBC URLs with “edb” do not try and use PG/EDB driver.**

* `liquibase log-level INFO update-sql --changelog-file <any changelog>`
* update-sql is successful
* there is no error about “pg_settings” in the output.

- - -
**Setup Case 2**

* Follow the instructions in [https://datical.atlassian.net/browse/LB-2185](https://datical.atlassian.net/browse/LB-2185|smart-link)  to start an EDB instance **on the default port (5444)**
* Create a database named ‘test’
* Update liquibase.properties to connect to the database:

` url:jdbc:edb://localhost:5444/test`


* Do not include driver name or driver jar in liquibase.properties file.
* Copy the edb jar into the liquibase/lib directory.
* Rename or delete the postgresql driver jar in liquibase/lib.

**Verify the EDB driver is used for a connection that has “edb” AND “:5444”**

* `liquibase log-level INFO update-sql --changelog-file <anychangelog>`
* update-sql is successful
* you do not get an error that the EDB driver cannot be located.

- - -
**Setup: Case 3**

* Start a second posgres  database on the default port (5432)
* Create a database named ‘test1’
* Update liquibase.properties to connect to the database:

`url: jdbc:postgresql://localhost:5432/test`


* Do not include driver name or driver jar in liquibase.properties file.
* Ensure the postgres driver is in liquibase/lib

**Verify the Postgres driver is used for a connection that has “postgres” and “:5432”**

* `liquibase log-level INFO update-sql --changelog-file <anychangelog>`
* update-sql is successful
* you do not get an error that the EDB driver cannot be located.
* the postgres driver is used.

- - -
**Automated Tests**

* None required for this ticket; there are nice unit tests for this change.



┆Issue is synchronized with this [Jira Bug](https://datical.atlassian.net/browse/LB-2214) by [Unito](https://www.unito.io)
